src/elliptic/curves/secp256_k1.rs

```diff
@@ -54,415 +74,251 @@ const BASE_POINT2_Y: [u8; 32] = [
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum Secp256k1 {}
+
+impl Curve for Secp256k1 {
+    type Point = GE;
+    type Scalar = FE;
+
+    fn curve_name() -> &'static str {
+        "secp256k1"
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Secp256k1Scalar {
+    purpose: &'static str,
+    /// Zeroizing<SK> wraps SK and zeroize it on drop
+    ///
+    /// `fe` might be None — special case for scalar being zero
+    fe: zeroize::Zeroizing<Option<SK>>,
+}
+#[derive(Clone, Debug, Copy)]
+pub struct Secp256k1Point {
+    purpose: &'static str,
+    ge: Option<PK>,
+}
+
+type GE = Secp256k1Point;
+type FE = Secp256k1Scalar;
+
 impl ECScalar for Secp256k1Scalar {
-    type SecretKey = SK;
+    type Underlying = Option<SK>;
 
-    fn new_random() -> Secp256k1Scalar {
+    fn random() -> Secp256k1Scalar {
+        let sk = SK(SecretKey::new(&mut thread_rng()));
         Secp256k1Scalar {
             purpose: "random",
-            fe: SecretKey::new(&mut thread_rng()),
+            fe: Zeroizing::new(Some(sk)),
         }
     }
 
     fn zero() -> Secp256k1Scalar {
-        let zero_arr = [0u8; 32];
-        let zero = unsafe { std::mem::transmute::<[u8; 32], SecretKey>(zero_arr) };
         Secp256k1Scalar {
             purpose: "zero",
-            fe: zero,
+            fe: Zeroizing::new(None),
         }
     }
 
-    fn get_element(&self) -> SK {
-        self.fe
+    fn is_zero(&self) -> bool {
+        self.fe.is_none()
     }
 
-    fn set_element(&mut self, element: SK) {
-        self.fe = element
-    }
+    fn from_bigint(n: &BigInt) -> Secp256k1Scalar {
+        if n.is_zero() {
+            return Self::zero();
+        }
 
-    fn from(n: &BigInt) -> Secp256k1Scalar {
-        let curve_order = FE::q();
-        let n_reduced = BigInt::mod_add(n, &BigInt::from(0), &curve_order);
-        let mut v = BigInt::to_bytes(&n_reduced);
+        let curve_order = Self::curve_order();
+        let n_reduced = n.modulus(curve_order);
+        let bytes = BigInt::to_bytes(&n_reduced);
 
-        if v.len() < SECRET_KEY_SIZE {
-            let mut template = vec![0; SECRET_KEY_SIZE - v.len()];
-            template.extend_from_slice(&v);
-            v = template;
-        }
+        let bytes = if bytes.len() < SECRET_KEY_SIZE {
+            let mut zero_prepended = vec![0; SECRET_KEY_SIZE - bytes.len()];
+            zero_prepended.extend_from_slice(&bytes);
+            zero_prepended
+        } else {
+            bytes
+        };
 
         Secp256k1Scalar {
-            purpose: "from_big_int",
-            fe: SK::from_slice(&v).unwrap(),
+            purpose: "from_bigint",
+            fe: Zeroizing::new(SecretKey::from_slice(&bytes).map(SK).ok()),
         }
     }
 
-    fn to_big_int(&self) -> BigInt {
-        BigInt::from_bytes(&(self.fe[0..self.fe.len()]))
-    }
-
-    fn q() -> BigInt {
-        BigInt::from_bytes(CURVE_ORDER.as_ref())
+    fn to_bigint(&self) -> BigInt {
+        match self.fe.deref() {
```

Comment:
nit, you can also do `match &*self.fe`, not sure if it's really better though

Reply:
I like it! [Fixed](https://github.com/ZenGo-X/curv/pull/120/commits/c50e75768092ebdb09c1278ecb591ea1ea076bd8)